fedmsg/crypto/x509_ng.py

```diff
@@ -19,256 +19,221 @@
-        **config)
-    crl = M2Crypto.X509.load_crl(crl)
-
-    # FIXME -- We need to check that the CRL is signed by our own CA.
-    # See https://bugzilla.osafoundation.org/show_bug.cgi?id=12954#c2
-    # if not ctx.validate_certificate(crl):
-    #    return fail("X509 CRL is not valid.")
-
-    # FIXME -- we check the CRL, but by doing string comparison ourselves.
-    # This is not what we want to be doing.
-    # There is a patch into M2Crypto to handle this for us.  We should use it
-    # once its integrated upstream.
-    # See https://bugzilla.osafoundation.org/show_bug.cgi?id=12954#c2
-    revoked_serials = [long(line.split(': ')[1].strip(), base=16)
-                       for line in crl.as_text().split('\n')
-                       if 'Serial Number:' in line]
-    if cert.get_serial_number() in revoked_serials:
-        subject = cert.get_subject()
-
-        signer = '(no CN)'
-        if subject.nid.get('CN'):
-            entry = subject.get_entries_by_nid(subject.nid['CN'])[0]
-            if entry:
-                signer = entry.get_data().as_text()
-
-        return fail("X509 cert %r, %r is in the Revocation List (CRL)" % (
-            signer, cert.get_serial_number()))
-
-    # If the cert is good, then test to see if the signature in the messages
-    # matches up with the provided cert.
-    rsa_public = cert.get_pubkey().get_rsa()
-    digest = M2Crypto.EVP.MessageDigest('sha1')
-    digest.update(fedmsg.encoding.dumps(message))
+        **config
+    )
+
+    with open(ca_file, 'rb') as fd:
+        ca_certificate = fd.read()
+
+    crl = None
+    if crl_file:
+        with open(crl_file, 'rb') as fd:
+            crl = fd.read()
+
     try:
-        if not rsa_public.verify(digest.digest(), signature):
-            raise M2Crypto.RSA.RSAError("RSA signature failed to validate.")
-    except M2Crypto.RSA.RSAError as e:
-        return fail(str(e))
+        _validate_signing_cert(ca_certificate, certificate, crl)
+    except X509StoreContextError as e:
+        _log.error(str(e))
+        return False
 
-    # Now we know that the cert is valid.  The message is *authenticated*.
-    # * Next step:  Authorization *
+    # Validate the signature of the message itself
+    try:
+        crypto_certificate = x509.load_pem_x509_certificate(certificate, default_backend())
+        crypto_certificate.public_key().verify(
+            signature,
+            fedmsg.encoding.dumps(message).encode('utf-8'),
+            asymmetric.padding.PKCS1v15(),
+            hashes.SHA1(),
+        )
+    except InvalidSignature as e:
+        _log.error('message [{m}] has an invalid signature: {e}'.format(
+            m=message, e=str(e)))
+        return False
 
-    # Load our policy from the config dict.
+    # Step 4, check that the certificate is permitted to emit messages for the
+    # topic.
+    common_name = crypto_certificate.subject.get_attributes_for_oid(x509.oid.NameOID.COMMON_NAME)
```

Comment:
You should not use the common name ideally, this is being deprecated by the CA/Browser Forum baseline requirements.
While we are not reliant on those, I think ideally we should follow their move to the Subject Alternative Name.
This does mean there may be multiple records, and we might need to check all SAN values in the authorization step.
For the time being, for compatibility, we might add the Common Name to the list of SANs for checking purposes though and issue a warning if the CN was not in any SAN values.